### PR TITLE
TAJO-1926: Disable partition pruning using catalog temporarily.

### PR DESCRIPTION
**Pull request by blrunner**
I reverted TAJO-1926 and added unit test cases.

### CODE REVIEW

File: tajo-catalog/tajo-catalog-server/src/test/java/org/apache/tajo/catalog/CatalogTestingUtil.java [hunk 1]
```diff
@@ -160,8 +160,8 @@ public static PartitionDesc buildPartitionDesc(String partitionName) {
       }
 
       PartitionKeyProto.Builder builder = PartitionKeyProto.newBuilder();
-      builder.setColumnName(columnName);
-      builder.setPartitionValue(partitionValue);
+      builder.setColumnName(partitionValue);
+      builder.setPartitionValue(columnName);
```
Comment (jihoonson) [inline tajo-catalog/tajo-catalog-server/src/test/java/org/apache/tajo/catalog/CatalogTestingUtil.java]: Should be exchanged.

File: tajo-catalog/tajo-catalog-drivers/tajo-hive/src/test/java/org/apache/tajo/catalog/store/TestHiveCatalogStore.java [hunk 2]
```diff
@@ -543,7 +405,7 @@ public void testTableUsingSequenceFileWithBinarySerde() throws Exception {
     }
 
     assertEquals(StorageConstants.DEFAULT_BINARY_SERDE,
-      table1.getMeta().getOption(StorageConstants.SEQUENCEFILE_SERDE));
+        table1.getMeta().getOption(StorageConstants.SEQUENCEFILE_SERDE));
```
Comment (jihoonson) [inline tajo-catalog/tajo-catalog-drivers/tajo-hive/src/test/java/org/apache/tajo/catalog/store/TestHiveCatalogStore.java]: Unnecessary change.

File: tajo-core-tests/src/test/java/org/apache/tajo/engine/query/TestTablePartitions.java [hunk 3]
```diff
@@ -433,15 +433,8 @@ public final void testColumnPartitionedTableByThreeColumns() throws Exception {
...
-    res = executeString("select * from " + tableName + " WHERE (col1 ='1' or col1 = '100') and col3 > 20");
-    String result = resultSetToString(res);
-    String expectedResult = "col4,col1,col2,col3\n" +
-      "-------------------------------\n" +
-      "N,1,1,36.0\n";
-    res.close();
-    assertEquals(expectedResult, result);
-
     res = executeString("SELECT col1, col2, col3 FROM " + tableName);
+    String result = resultSetToString(res);
```
Comment (jihoonson) [inline tajo-core-tests/src/test/java/org/apache/tajo/engine/query/TestTablePartitions.java]: Unused variable

File: tajo-core-tests/src/test/java/org/apache/tajo/engine/query/TestTablePartitions.java [hunk 4]
```diff
@@ -1340,411 +1327,12 @@ public final void testDuplicatedPartitions() throws Exception {
...
-    res.close();
-    assertEquals(expectedResult, result);
-
-    executeString("DROP TABLE " + databaseName + "." + tableName + " PURGE").close();
-    executeString("DROP database " + databaseName).close();
-  }
-
-  @Test
-  public void testAbnormalDirectories()  throws Exception {
+  public void testExternalTable()  throws Exception {
```
Comment (jihoonson) [inline tajo-core-tests/src/test/java/org/apache/tajo/engine/query/TestTablePartitions.java]: This test includes testing on managed tables as well as external tables. Please change the name.

File: tajo-catalog/tajo-catalog-drivers/tajo-hive/src/main/java/org/apache/tajo/catalog/store/HiveCatalogStore.java [hunk 5]
```diff
@@ -1243,7 +1031,7 @@ public void addPartitions(String databaseName, String tableName, List<CatalogPro
         // Unfortunately, hive client add_partitions doesn't run as expected. The method never read the ifNotExists
         // parameter. So, if Tajo adds existing partition to Hive, it will threw AlreadyExistsException. To avoid
         // above error, we need to filter existing partitions before call add_partitions.
-        if (existingPartition == null) {
+        if (existingPartition != null) {
```
Comment (jihoonson) [inline tajo-catalog/tajo-catalog-drivers/tajo-hive/src/main/java/org/apache/tajo/catalog/store/HiveCatalogStore.java]: This change is wrong.

**Comment by blrunner** [comment]
@jihoonson 

Thank you for your review. I reflected your comments and added unit test cases for PartitionedTableRewriter::findFilteredPaths

File: tajo-plan/src/main/java/org/apache/tajo/plan/rewrite/rules/PartitionedTableRewriter.java [hunk 6]
```diff
@@ -421,6 +315,25 @@ private boolean checkIfDisjunctiveButOneVariable(EvalNode evalNode) {
...
+      throws TajoException {
+    if (scanNode.getInputPaths().length > 0) {
+      try {
+        FileSystem fs = scanNode.getInputPaths()[0].getFileSystem(queryContext.getConf());
+        long totalVolume = 0;
+
+        for (Path input : scanNode.getInputPaths()) {
+          ContentSummary summary = fs.getContentSummary(input);
+          totalVolume += summary.getLength();
+          totalVolume += summary.getFileCount();
```
Comment (jihoonson) [inline tajo-plan/src/main/java/org/apache/tajo/plan/rewrite/rules/PartitionedTableRewriter.java]: File count must not be added.

File: tajo-storage/tajo-storage-hdfs/src/main/java/org/apache/tajo/storage/FileTablespace.java [hunk 7]
```diff
@@ -343,9 +343,9 @@ public boolean accept(Path path) {
 
       FileStatus[] matches = fs.globStatus(p, inputFilter);
       if (matches == null) {
-        LOG.warn("Input path does not exist: " + p);
+        errors.add(new IOException("Input path does not exist: " + p));
       } else if (matches.length == 0) {
-        LOG.warn("Input Pattern " + p + " matches 0 files");
+        errors.add(new IOException("Input Pattern " + p + " matches 0 files"));
```
Comment (jihoonson) [inline tajo-storage/tajo-storage-hdfs/src/main/java/org/apache/tajo/storage/FileTablespace.java]: The above two cases are not errors.

File: tajo-plan/src/main/java/org/apache/tajo/plan/rewrite/rules/PartitionedTableRewriter.java (hunk 6 above)
Comment (blrunner) [inline tajo-plan/src/main/java/org/apache/tajo/plan/rewrite/rules/PartitionedTableRewriter.java]: I removed above code.

File: tajo-storage/tajo-storage-hdfs/src/main/java/org/apache/tajo/storage/FileTablespace.java (hunk 7 above)
Comment (blrunner) [inline tajo-storage/tajo-storage-hdfs/src/main/java/org/apache/tajo/storage/FileTablespace.java]: I applied existing codes. :)

File: tajo-cli/src/main/java/org/apache/tajo/cli/tools/TajoDump.java [hunk 8]
```diff
@@ -186,6 +186,7 @@ private static void dumpDatabase(TajoClient client, String databaseName, PrintWr
         } else {
           writer.write(DDLBuilder.buildDDLForBaseTable(table));
         }
+
         if (table.hasPartition()) {
```
Comment (jihoonson) [inline tajo-cli/src/main/java/org/apache/tajo/cli/tools/TajoDump.java]: Since we don't support partition dump yet, the whole if block should be removed or commented out. 
In addition, you need to print an warning message for users to know that partition dump and restore are not supported yet.

File: tajo-core-tests/src/test/java/org/apache/tajo/engine/planner/TestPartitionedTableRewriter.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+import java.sql.ResultSet;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+
+public class TestPartitionedTableRewriter extends QueryTestCaseBase {
+  @Before
+  public void setUp() throws Exception {
+    executeString("create database " + getCurrentDatabase()).close();
```
Comment (jihoonson) [inline tajo-core-tests/src/test/java/org/apache/tajo/engine/planner/TestPartitionedTableRewriter.java]: Current database is automatically created and dropped. You don't need to do explicitly.

File: tajo-core-tests/src/test/java/org/apache/tajo/engine/planner/TestPartitionedTableRewriter.java [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+    TableDesc tableDesc = catalog.getTableDesc(databaseName, tableName);
+    assertNotNull(tableDesc);
+
+    PartitionedTableRewriter rewriter = new PartitionedTableRewriter();
+    OverridableConf conf = CommonTestingUtil.getSessionVarsForTest();
+
+    Schema partitionColumns = tableDesc.getPartitionMethod().getExpressionSchema();
+
+    // Get all partitions
+    Path[] filteredPaths = rewriter.findFilteredPaths(conf, partitionColumns, null, new Path(tableDesc.getUri()));
```
Comment (jihoonson) [inline tajo-core-tests/src/test/java/org/apache/tajo/engine/planner/TestPartitionedTableRewriter.java]: You should test ```PartitionedTableRewriter::findFilteredPartitionPaths()``` instead of ```PartitionedTableRewriter::findFilteredPaths()```.

File: tajo-core-tests/src/test/java/org/apache/tajo/engine/query/TestTablePartitions.java [hunk 11]
```diff
@@ -596,7 +588,7 @@ public final void testInsertIntoColumnPartitionedTableByThreeColumns() throws Ex
   }
 
   private final void verifyKeptExistingData(ResultSet res, String tableName) throws Exception {
-    res = executeString("select * from " + tableName + " where col2 = 1 order by col4, col1, col2, col3");
+    res = executeString("select * from " + tableName + " where col2 = 1");
```
Comment (jihoonson) [inline tajo-core-tests/src/test/java/org/apache/tajo/engine/query/TestTablePartitions.java]: ```order by``` clause must be added.

File: tajo-core/src/main/java/org/apache/tajo/engine/planner/physical/ColPartitionStoreExec.java [hunk 12]
```diff
@@ -174,18 +173,17 @@ private void addPartition(String partition) throws IOException {
 
       PartitionKeyProto.Builder keyBuilder = PartitionKeyProto.newBuilder();
       keyBuilder.setColumnName(split[0]);
-      // Partition path have been escaped to avoid URISyntaxException. But partition value of partition keys table
-      // need to contain unescaped value for comparing filter conditions in select statement.
-      keyBuilder.setPartitionValue(StringUtils.unescapePathName(split[1]));
```
Comment (jihoonson) [inline tajo-core/src/main/java/org/apache/tajo/engine/planner/physical/ColPartitionStoreExec.java]: This code must be added.

File: tajo-core/src/main/java/org/apache/tajo/engine/planner/physical/ColPartitionStoreExec.java [hunk 13]
```diff
@@ -174,18 +173,17 @@ private void addPartition(String partition) throws IOException {
...
+      keyBuilder.setPartitionValue(split[1]);
 
       builder.addPartitionKeys(keyBuilder.build());
     }
 
     if (this.plan.getUri() == null) {
-      // In CTAS, the uri would be null. So, it get the uri from staging directory.
-      int endIndex = storeTablePath.toString().indexOf(FileTablespace.TMP_STAGING_DIR_PREFIX);
```
Comment (jihoonson) [inline tajo-core/src/main/java/org/apache/tajo/engine/planner/physical/ColPartitionStoreExec.java]: This code must be added.

**Comment by jihoonson** [comment]
This patch cannot pass unit tests.

```
Results :

Failed tests:
  TestDDLBuilder.testBuildDDLForExternalTable:64 expected:<...) USING TEXT WITH ('[compression.codec'='org.apache.hadoop.io.compress.GzipCodec', 'text.delimiter'='|]') PARTITION BY COLU...> but was:<...) USING TEXT WITH ('[text.delimiter'='|', 'compression.codec'='org.apache.hadoop.io.compress.GzipCodec]') PARTITION BY COLU...>
  TestDDLBuilder.testBuildDDLForBaseTable:103 expected:<...) USING TEXT WITH ('[compression.codec'='org.apache.hadoop.io.compress.GzipCodec', 'text.delimiter'='|]');> but was:<...) USING TEXT WITH ('[text.delimiter'='|', 'compression.codec'='org.apache.hadoop.io.compress.GzipCodec]');>
  TestDDLBuilder.testBuildDDLQuotedTableName:90 expected:<...) USING TEXT WITH ('[compression.codec'='org.apache.hadoop.io.compress.GzipCodec', 'text.delimiter'='|]') PARTITION BY COLU...> but was:<...) USING TEXT WITH ('[text.delimiter'='|', 'compression.codec'='org.apache.hadoop.io.compress.GzipCodec]') PARTITION BY COLU...>

Tests run: 1692, Failures: 3, Errors: 0, Skipped: 0
```

**Comment by blrunner** [comment]
@jihoonson 

Thanks for your detailed review. I reflected your comments and fixed unit test bugs for DDLBuilder.

File: tajo-cli/src/main/java/org/apache/tajo/cli/tools/TajoDump.java [hunk 14]
```diff
@@ -216,6 +217,8 @@ private static void dumpDatabase(TajoClient client, String databaseName, PrintWr
         System.err.println("ERROR:" + tableName + "," + e.getMessage());
       }
     }
+
+    LOG.warn("Partition dump and restore are not supported yet");
```
Comment (jihoonson) [inline tajo-cli/src/main/java/org/apache/tajo/cli/tools/TajoDump.java]: Where is this message printed? Have you ever tested by yourself?
In addition, this message should be printed only when users try to dump partitioned table.

Comment (blrunner) [inline tajo-cli/src/main/java/org/apache/tajo/cli/tools/TajoDump.java]: I had misunderstood your comments and added a printed message and unit test cases.

File: tajo-core-tests/src/test/java/org/apache/tajo/engine/planner/TestPartitionedTableRewriter.java [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+import org.apache.tajo.plan.logical.*;
+import org.apache.tajo.plan.rewrite.rules.PartitionedTableRewriter;
+import org.apache.tajo.util.CommonTestingUtil;
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+public class TestPartitionedTableRewriter extends QueryTestCaseBase {
```
Comment (jihoonson) [inline tajo-core-tests/src/test/java/org/apache/tajo/engine/planner/TestPartitionedTableRewriter.java]: You must add a test to check that findFilteredPartitionPaths() works well after a manual partition directory creation. This test is related to the following bug.

```
# I manually added a directory 'n_comment=this is a directory to test partition' under nation_part directory.
default> \dfs -ls /tajo/warehouse/default/nation_part
Found 26 items
drwxr-xr-x   - jihoon supergroup          0 2015-10-15 13:03 /tajo/warehouse/default/nation_part/n_comment= haggle. carefully final deposits detect slyly agai
drwxr-xr-x   - jihoon supergroup          0 2015-10-15 13:03 /tajo/warehouse/default/nation_part/n_comment= pending excuses haggle furiously deposits. pending, express pinto beans wake fluffily past t
drwxr-xr-x   - jihoon supergroup          0 2015-10-15 13:03 /tajo/warehouse/default/nation_part/n_comment= requests against the platelets use never according to the quickly regular pint
drwxr-xr-x   - jihoon supergroup          0 2015-10-15 13:03 /tajo/warehouse/default/nation_part/n_comment= slyly express asymptotes. regular deposits haggle slyly. carefully ironic hockey players sleep blithely. carefull
drwxr-xr-x   - jihoon supergroup          0 2015-10-15 13:03 /tajo/warehouse/default/nation_part/n_comment=al foxes promise slyly according to the regular accounts. bold requests alon
drwxr-xr-x   - jihoon supergroup          0 2015-10-15 13:03 /tajo/warehouse/default/nation_part/n_comment=c dependencies. furiously express notornis sleep slyly regular accounts. ideas sleep. depos
drwxr-xr-x   - jihoon supergroup          0 2015-10-15 13:03 /tajo/warehouse/default/nation_part/n_comment=eans boost carefully special requests. accounts are. carefull
drwxr-xr-x   - jihoon supergroup          0 2015-10-15 13:03 /tajo/warehouse/default/nation_part/n_comment=eas hang ironic, silent packages. slyly regular packages are furiously over the tithes. fluffily bold
drwxr-xr-x   - jihoon supergroup          0 2015-10-15 13:03 /tajo/warehouse/default/nation_part/n_comment=efully alongside of the slyly final dependencies. 
drwxr-xr-x   - jihoon supergroup          0 2015-10-15 13:03 /tajo/warehouse/default/nation_part/n_comment=hely enticingly express accounts. even, final 
drwxr-xr-x   - jihoon supergroup          0 2015-10-15 13:03 /tajo/warehouse/default/nation_part/n_comment=ic deposits are blithely about the carefully regular pa
drwxr-xr-x   - jihoon supergroup          0 2015-10-15 13:03 /tajo/warehouse/default/nation_part/n_comment=l platelets. regular accounts x-ray%3A unusual, regular acco
drwxr-xr-x   - jihoon supergroup          0 2015-10-15 13:03 /tajo/warehouse/default/nation_part/n_comment=nic deposits boost atop the quickly final requests%3F quickly regula
drwxr-xr-x   - jihoon supergroup          0 2015-10-15 13:03 /tajo/warehouse/default/nation_part/n_comment=ously. final, express gifts cajole a
drwxr-xr-x   - jihoon supergroup          0 2015-10-15 13:03 /tajo/warehouse/default/nation_part/n_comment=platelets. blithely pending dependencies use fluffily across the even pinto beans. carefully silent accoun
drwxr-xr-x   - jihoon supergroup          0 2015-10-15 13:03 /tajo/warehouse/default/nation_part/n_comment=refully final requests. regular, ironi
drwxr-xr-x   - jihoon supergroup          0 2015-10-15 13:03 /tajo/warehouse/default/nation_part/n_comment=rns. blithely bold courts among the closely regular packages use furiously bold platelets%3F
drwxr-xr-x   - jihoon supergroup          0 2015-10-15 13:03 /tajo/warehouse/default/nation_part/n_comment=s. ironic, unusual asymptotes wake blithely r
drwxr-xr-x   - jihoon supergroup          0 2015-10-15 13:03 /tajo/warehouse/default/nation_part/n_comment=ss excuses cajole slyly across the packages. deposits print aroun
drwxr-xr-x   - jihoon supergroup          0 2015-10-15 17:45 /tajo/warehouse/default/nation_part/n_comment=this is a directory to test partition.
drwxr-xr-x   - jihoon supergroup          0 2015-10-15 13:03 /tajo/warehouse/default/nation_part/n_comment=ts. silent requests haggle. closely express packages sleep across the blithely
drwxr-xr-x   - jihoon supergroup          0 2015-10-15 13:03 /tajo/warehouse/default/nation_part/n_comment=ular asymptotes are about the furious multipliers. express dependencies nag above the ironically ironic account
drwxr-xr-x   - jihoon supergroup          0 2015-10-15 13:03 /tajo/warehouse/default/nation_part/n_comment=ven packages wake quickly. regu
drwxr-xr-x   - jihoon supergroup          0 2015-10-15 13:03 /tajo/warehouse/default/nation_part/n_comment=y above the carefully unusual theodolites. final dugouts are quickly across the furiously regular d
drwxr-xr-x   - jihoon supergroup          0 2015-10-15 13:03 /tajo/warehouse/default/nation_part/n_comment=y alongside of the pending deposits. carefully special packages are about the ironic forges. slyly special 
drwxr-xr-x   - jihoon supergroup          0 2015-10-15 13:03 /tajo/warehouse/default/nation_part/n_comment=y final packages. slow foxes cajole quickly. quickly silent platelets breach ironic accounts. unusual pinto be
default> select n_comment from nation_part where substr(n_comment, 0, 8) = 'this is';
Progress: 100%, response time: 2.031 sec
n_comment
-------------------------------
(0 rows, 2.031 sec, 0 B selected)
```

Comment (blrunner) [inline tajo-core-tests/src/test/java/org/apache/tajo/engine/planner/TestPartitionedTableRewriter.java]: I added unit test cases for a manual partition directory creation. And if users make directories with empty string like above cases, Tajo can't scan those directories because PartitionPathFilter can't check abnormal directories. I think that above problem need to handle at another issue.

File: tajo-core-tests/src/test/java/org/apache/tajo/engine/planner/TestPartitionedTableRewriter.java [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+    // Get all partitions
+    verifyFindFilteredPartitionPaths("SELECT * FROM " + canonicalTableName + " ORDER BY key", 1);
+
+    // Get partition with filter condition
+    verifyFindFilteredPartitionPaths("SELECT * FROM " + canonicalTableName + " WHERE key = 17.0 ORDER BY key", 2);
+
+    executeString("DROP TABLE " + canonicalTableName + " PURGE").close();
+  }
+
+  private void verifyFindFilteredPartitionPaths(String sql, int type) throws Exception {
```
Comment (jihoonson) [inline tajo-core-tests/src/test/java/org/apache/tajo/engine/planner/TestPartitionedTableRewriter.java]: ```type``` is not defined in anywhere, so it is difficult to understand.
This kind of coding style is very poor and we must avoid. I cannot understand why you try to verify results of different queries in a single method. 
If you want to avoid repeating same code blocks, you should create a method only for those repetitive ones.

File: tajo-core-tests/src/test/java/org/apache/tajo/engine/planner/TestPartitionedTableRewriter.java (hunk 15 above)
Comment (jihoonson) [inline tajo-core-tests/src/test/java/org/apache/tajo/engine/planner/TestPartitionedTableRewriter.java]: I cannot understand what you mean. I didn't make a directory of empty string.
Anyway, this is the case when the partition directory doesn't contain any data file. So this is not a problem.
After adding a data file, everything works well.

File: tajo-core-tests/src/test/java/org/apache/tajo/engine/planner/TestPartitionedTableRewriter.java (hunk 16 above)
Comment (blrunner) [inline tajo-core-tests/src/test/java/org/apache/tajo/engine/planner/TestPartitionedTableRewriter.java]: Reflected your comments.

**Comment by jihoonson** [comment]
You need to add more tests to TestPartitionedTableRewriter.
At least the following cases should be tested.
* Without any filters
* With a filter which checks non-existing partition value
* With a filter which checks non-partition key columns
* Given a partition key of multiple columns,
 * The case when a filter involves every partition key column
 * The case when a filter involves some of partition key columns
 * The case when a filter involves non-partition key columns

**Comment by blrunner** [comment]
Thanks @jihoonson . I've added more unit test cases.

File: tajo-core-tests/src/test/java/org/apache/tajo/engine/planner/TestPartitionedTableRewriter.java [hunk 17]
```diff
@@ -0,0 +1,394 @@
...
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+public class TestPartitionedTableRewriter extends QueryTestCaseBase {
+
+  @Test
+  public final void testPartitionPruningWithManualPartitionPathCreation() throws Exception {
```
Comment (jihoonson) [inline tajo-core-tests/src/test/java/org/apache/tajo/engine/planner/TestPartitionedTableRewriter.java]: This is not an idiomatic way for junit testing. 
Usually, ```setup()``` and ```teardown()``` methods are used for preparing and finalizing each test. With this way, it is easier to exactly figure out which functions don't work.
So, you need to move the preparing and finalizing codes to ```setup()``` and ```teardown()``` methods and make the below methods of which name starts with ```test``` as separate test methods.

Comment (blrunner) [inline tajo-core-tests/src/test/java/org/apache/tajo/engine/planner/TestPartitionedTableRewriter.java]: Reflected your comments.

**Comment by jihoonson** [comment]
+1 LGTM
